### ISSUE TITLE
A bug about slope function

### ISSUE THREAD

**Issue by rdzzz** [lammps/lammps#662]
The keyword “slope” is a math function in variable command. However, I found a bug that this function calculated wrong answer even for the simplest vector (1, 3, 5).
Here is my test input file.
units		lj
atom_style	atomic

lattice		fcc 0.8442
region		box block 0 10 0 10 0 10
create_box	1 box
create_atoms	1 box
mass		1 1.0

velocity	all create 3.0 87287

pair_style	lj/cut 2.5
pair_coeff	1 1 1.0 1.0 2.5

neighbor	0.3 bin
neigh_modify	every 20 delay 0 check no

fix		1 all nve
**variable        a equal 1+2*step
fix             2 all vector 1 v_a
variable        k equal slope(f_2)**
thermo		1
thermo_style    custom step v_a v_k 
run		2

And part of the output information is here.
Step v_a v_k 
       0            1       1e+020 
       1            3       1e+020 
       2            5    2.3333333 

When timestep is 2, the fix 2 generates a vector (1, 3, 5). However, the output information shows that the slope of the vector (variable k) is 2.333, which is obviously wrong.
I read the sourcecode in the file variable.cpp and I rewrote relative code to a python code (see attachment file)
[testSlopeofPython.txt](https://github.com/lammps/lammps/files/1328499/testSlopeofPython.txt)

I analyse the code and find that the code is following the equation below:
slope of Lammps = (n-1)/n * [sum(xi * yi) - sum(xi) * sum(yi)] / [sum(xi**2) - (sum(xi))**2]
This equation is wrong. 

The correct equation of least square method is
[n * sum(xi * yi) - sum(xi) * sum(yi)] / [n * sum(xi**2) - sum(xi) * sum(xi)]

**Comment by akohlmey** [comment]
@sjplimp i think this is something you should look at.

**Comment by rdzzz** [comment]
Have anybody seen this issue? I think this is a bug that lasts many years.

**Comment by akohlmey** [comment]
@rdzzz yes, this is being monitored. however, it is not a high-priority item, since the impact for typical use cases is negligible, since the difference between the two expression becomes tiny after applying it to several hundred data points.

**Comment by akohlmey** [comment]
@sjplimp sending a reminder. this should be simple to address, as it looks like an off-by-one error, that is only relevant for a small number of data items, which is not a typical use case.

**Comment by quang-ha** [comment]
I can tackle this problem since this looks simple and straightforward. @akohlmey is there a code guideline/testing I need to follow before making the pull request?

**Comment by akohlmey** [comment]
@quang-ha the fix should be straightforward. you should compare the results in the DIFFUSE example codes. first test the exiting data against an unmodified executable, then with the modification. the difference should be very, very small. then test with the test example in the first message in this issue. please note, that this test is a corner case, and the error should diminish as more data points are considered. also, you should coordinate with @sjplimp. this is a rather sensitive piece of code, so any mistake or change with unexpected side effects can negatively affect many LAMMPS users.

**Comment by quang-ha** [comment]
Using the updated linear regression formula, I could match the results here, but only one case in DIFFUSE match - msd match pretty closely but vacf swayed further away (0.68 instead of expected 0.25-0.45)...

**Comment by quang-ha** [comment]
@sjplimp any comments? Using the old formula I will get the result range for both DIFFUSE test, but of course will miss out the simple test case here. Using the newly obtained least square equation will give correct answer here, but the 2nd case for DIFFUSE would be incorrect.

**Comment by sjplimp** [comment]
@rdzzz I agree with your formula correction.  Not sure how this happened.  Possibly I
was trying to be too clever with rescaling the x values to be from 0 to 1 (to avoid
round-off), and messed up the formula.  In the new version, I just accumulate
the x and x^2 values in a bigint, so it is simpler now.  Thanks for catching this.
It is a small correction for large N values in the DIFFUSE problem, but I'm not
clear that would be the case for any set of y values.

@quang-ha You can check if this is giving the same answers as your version.
I don't understand the comment about VACF, since the DIFFUSE/in.vacf script
does not use slope()

**Comment by sjplimp** [comment]
Meant to add I have submitted a PR for this change, so it should be in the next
patch and stable release.

**Comment by quang-ha** [comment]
I was just skeptical with my attempt (https://github.com/quang-ha/lammps/tree/quang/fix-662). In the README of DIFFUSE it says the vacf should have value of 0.25-0.45. I keep getting 0.6ish in my code so...
